packages/reactivity/src/effect.ts

```diff
@@ -5,38 +5,114 @@ import { EMPTY_OBJ, isArray, isIntegerKey, isMap } from '@vue/shared'
 // Conceptually, it's easier to think of a dependency as a Dep class
 // which maintains a Set of subscribers, but we simply store them as
 // raw Sets to reduce memory overhead.
+//
+// Notice that refs store their deps in a local property for
+// performance reasons.
 type Dep = Set<ReactiveEffect>
 type KeyToDepMap = Map<any, Dep>
 const targetMap = new WeakMap<any, KeyToDepMap>()
 
-export interface ReactiveEffect<T = any> {
-  (): T
-  _isEffect: true
-  id: number
+type EffectScheduler = (job: () => void) => void
+
+export class ReactiveEffect<T = any> {
+  public id = uid++
+  public deps: Dep[] = []
+  private runner?: ReactiveEffectFunction<T>
+
+  constructor(
+    public raw: () => T,
+    allowRecurse: boolean,
+    public scheduler: EffectScheduler | undefined,
+    options: ReactiveEffectOptions | undefined
+  ) {
+    if (allowRecurse) {
+      this.allowRecurse = true
+    }
+    if (options) {
+      this.options = options
+    }
+  }
+
+  public setOnStop(func: () => void) {
+    if (this.options === EMPTY_OBJ) {
+      this.options = {}
+    }
+    this.options.onStop = func
+  }
+
+  public run() {
+    if (!this.active) {
+      return this.scheduler ? undefined : this.raw()
+    }
+    if (!effectStack.includes(this)) {
+      cleanup(this)
+      try {
+        enableTracking()
+        effectStack.push(this)
+        activeEffect = this
+        return this.raw()
+      } finally {
+        effectStack.pop()
+        resetTracking()
+        const n = effectStack.length
+        activeEffect = n ? effectStack[n - 1] : undefined
+      }
+    }
+  }
+
+  public get func(): ReactiveEffectFunction<T> {
+    if (!this.runner) {
+      const runner = () => {
+        return this.run()
+      }
+      runner.effect = this
+      runner.allowRecurse = this.allowRecurse
+      this.runner = runner
+    }
+    return this.runner
+  }
+}
+
+// Use prototype for optional properties to minimize memory usage.
+export interface ReactiveEffect {
   active: boolean
-  raw: () => T
-  deps: Array<Dep>
+  allowRecurse: boolean
   options: ReactiveEffectOptions
+}
+
+ReactiveEffect.prototype.active = true
+ReactiveEffect.prototype.allowRecurse = false
+ReactiveEffect.prototype.options = EMPTY_OBJ
```

Comment:
@basvanmeurs The code might be cleaner if you declare those three as getters on the class?
`public get active() { return true }` would end up on the prototype and I don't think usage of a getter rather than field would impact perf / memory in any noticeable way (we have your benchmarks to verify).

Other than coding style, it's also easier for tools to determine that a module is side-effect free if it doesn't contain statements, just class declarations.

Reply:
I was initially enthousiastic. But when I tried changing this, all property assignments started to cause typescript errors (Cannot assign to 'options' because it is a read-only property.). I could ignore them for every assignment by casting to any: `(this as any).options = {}`. But personally I think that the resulting code will be worse than what it is now.

Reply:
Ah, I did not see that one coming.
I agree your original code is a lot better than using `any` in callers.
I see no way to "idiomatically" have TS infer that we're having fields with default values stored in prototype.

If we did things in a "less-JS" way we would either mark those fields optional and provide the default value in code that reads them, or always init fields with the default value.
Not sure about the perf/memory impact of those ways, it would need some benchmarking.

Always assigning a value to a field member would increase memory usage but _might_ lead to monomorphic code, which _maybe_ executes faster. Not sure if it would be noticeable. You have a benchmark avail., maybe you could try.

Reply:
I tried setting the properties in the constructor during development for the reason you mentioned. It did not affect performance at all, at least in Chromium. It might be different in other browsers, but probably not by much. 

Because it really saved some memory I decided to put them on the prototype.